README.md
Project 1 - IPK calculator protocol
Author: Milan Takac - xtakac09
2BIT VUT FIT (BUT)

The goal of this project was to code a client, that will be able to send UDP/TCP request to server 
with Arithmetics-as-a-Service (Aaas) with simple equation (+ - * /), receive server's response with 
result and print it to terminal.

The client is started using: command make to compile the source code and get an executable file, then
using command "ipkcpc -h < host > -p < port > -m < mode >" where host is the IPv4 address of the server, 
port the server port, and mode either tcp or udp (e.g., ipkcpc -h 1.2.3.4 -p 2023 -m udp), then you see
SERVER INFO with ip address and port you are sending your requests to. To send a request, click on new 
line under SERVER INFO, insert your message and press ENTER. 

For UDP request send message such as: 
(+ 1 2)
(* 3 (- 4 2))
etc.
To end the programme, write "exit".

For TCP request you first have to send message "HELLO" to establish connection. then you can send messages
such as:
SOLVE (+ 1 2)
SOLVE (* 3 (- 4 2))
etc.
To end the programme, write "BYE".

If you terminate the process by pressing Ctrl + c, you will disconnect from the server, and with mode TCP, 
the connection will end as if you wrote "BYE".

File called ipkcp.c is kept for commit history, because in the process of development this was the 
file I was working on before renaming it to name specified by the assignment.

CHANGELOG.md
Project 1 - IPK calculator protocol
Author: Milan Takac - xtakac09
2BIT VUT FIT (BUT)

After compilation, an executable file is created. Using is described in README.

After executing the command, the code will:
- check set parameters
- translate the adress
- connect to server (TCP)
- read user's input from command line
- process the input (UDP): 
    - empty the buffer
    - set opcode to request
    - set payload length
    - add inserted message
    into the buffer
- send request
- empty the buffer
- receive response from server
- print the response
- read user's input from command line in case of repeated queries
- safely disconnect from server (TCP)

In case of caught interrupt signal (C-c), it terminates the communication (UDP). If you're 
communicating using TCP, there is a signal handler that sets variable keepRunning to 0, and
from waiting for user's input it escapes after the while loop, and safely disconnects from 
server.

Code is well documented, so there shouldn't be any problem with trying to understand it.

Different operating systems
There are different things to look for in code like this to be multiplatform. This code is 
written for UNIX systems. 
Few key differences are sockets. Each system uses different library for them. For example 
WINDOWS uses library winsock.h meanwile UNIX uses sys/socket.h. The process of creating them 
is also different. Another thing to take in mind are different byte orders. WINDOWS uses 
little-endian and UNIX uses big-endian. Unfortunately this code isn't implemented for using 
on WINDOWS.



ipkcp.c
#include <stdio.h>
#include <string.h>
#include <stdlib.h>         // atoi
#include <sys/socket.h>     // socket, sendto
#include <sys/types.h>
#include <netdb.h>          // gethostbyname
#include <arpa/inet.h>      // htons
#include <unistd.h>         // close
#include <signal.h>         // C-c interrupt signal

#define BUFFER_SIZE_UDP 512     
#define BUFFER_SIZE_TCP 1024     

// rozdil mezi win a unix bude mozna v adresach neceho, struktura sockaddr_in nebo tak

static volatile int keepRunning = 1;                                                    // Help variable in case of interrupt

void intHandler() {
    keepRunning = 0;
    return;
}

int main(int argc, char** argv){
    signal(SIGINT, intHandler);                                                         // For C-c interrupt

// Parameter check 
    if(argc != 7){                                                                      // Number of args
        fprintf(stderr, "ERROR: Wrong number of arguments. %d of 6 needed.\n", argc);
        return 1;
    }

    if(strcmp(argv[1], "-h") != 0 && strcmp(argv[3], "-p") != 0 && strcmp(argv[5], "-m") != 0){ 
        fprintf(stderr, "ERROR: Wrong arguments. Correct form is \"ipkcpc -h <host> -p <port> -m <mode>\"\n");
        return 1;
    }

    if(!(strcmp(argv[6], "tcp") == 0 || strcmp(argv[6], "udp") == 0)){
        fprintf(stderr, "ERROR: Wrong mode \n%s\n - choose \"tcp\" or \"udp\".\n", argv[6]);
        return 1;
    }

    if(atoi(argv[4]) < 0){                                                              // Checks if port number isn't negative
        fprintf(stderr, "ERROR: Wrong port parameter: %s.\n", argv[4]);
        return 1;
    }

    char conType[4];                                                                    // Connection type (UDP/TCP)
    strcpy(conType, argv[6]);

    char host_address[16];               
    strcpy(host_address, argv[2]);

    int port_number = atoi(argv[4]);
    char buffer_udp[BUFFER_SIZE_UDP];
    char buffer_tcp[BUFFER_SIZE_TCP];
    char input[BUFFER_SIZE_UDP];
    char input_tcp[BUFFER_SIZE_TCP];
    int family = AF_INET;                                                               // ipv4

    // Getting the message to send, getting server info 
    struct hostent *server = gethostbyname(host_address);                               // Gets info about server. Parameter takes either address or name
    if(server == NULL){                                                                 // Invalid server name
        fprintf(stderr, "ERROR: no such host %s.\n", host_address);
        return 1;
    }

    struct sockaddr_in server_address;
    memset(&server_address, 0, sizeof(server_address));                                 // fills server address with zeros

    server_address.sin_family = AF_INET;
    server_address.sin_port = htons(port_number);                                       // sets port that the socket will use. htonl() translates an unsigned long integer into network byte order
    memcpy(&server_address.sin_addr.s_addr, server->h_addr_list[0], server->h_length); 

    printf("INFO: Server socket: %s : %d \n", 
            inet_ntoa(server_address.sin_addr),                                         // server's in address
            ntohs(server_address.sin_port));                                            // server's in port

    socklen_t server_size = sizeof(server_address);    

// UDP ------------------------------------------------------------------------------------------
    if(strcmp(conType, "udp") == 0){                                                    // UDP - no need to disconnect if an error occurs
        int type = SOCK_DGRAM;

    // SOCKET() - creating socket
        int client_socket = socket(family, type, 0);
        if(client_socket <= 0){     
            fprintf(stderr, "ERROR: socket.\n");
            return 1;
        }

        while(keepRunning){
        // MESSAGE
            memset(buffer_udp, 0, BUFFER_SIZE_UDP);                                     // Fills buffer with 0
            
            if(fgets(input, BUFFER_SIZE_UDP, stdin) == NULL){                           // Reads from input 
                fprintf(stderr, "ERROR occured while reading input.\n");
                return 1;
            }

            if(!keepRunning){
                break;
            }

            if(strcmp(input, "exit\n") == 0){                                           // Exit message
                printf("Exiting...\n");
                return 0;
            }

            int inputLength = (int) strlen(input) - 1;                                  // - 1, because '\n' is not considered as part of the payload length
            buffer_udp[0] = 0;                                                          // Opcode (0 = request)
            buffer_udp[1] = inputLength;                                                // Payload Length

            for(int i  = 0; i < inputLength; i++){                                      // Fills the payload data area with user input
                buffer_udp[i + 2] = input[i];                                      
            }

        // SENDTO() - sending message to server 
            int bytes_tx = sendto(client_socket, buffer_udp, (inputLength + 2), 0, (struct sockaddr *) &server_address, server_size);    // inputLength + 2, because opcode + payloadLength + input (not counting '\0')
            if(bytes_tx < 0){
                fprintf(stderr, "ERROR: sendto.\n");
            }

            memset(buffer_udp, 0, sizeof(buffer_udp));                                  // Fills buffer with 0
            
        // RECVFROM() - waiting for response
            
            int bytes_rx = recvfrom(client_socket, buffer_udp, BUFFER_SIZE_UDP, 0, (struct sockaddr *) &server_address, &server_size);
            if(bytes_rx < 0){
                fprintf(stderr, "ERROR: recvfrom.\n");
                return 1;
            } 

            int responseStatus = buffer_udp[1];
            int responseLength = buffer_udp[3];
            if(responseStatus == 1){           // STATUS CODE: 0 = OK, 1 = ERR
                fprintf(stderr, "ERR:");
                for(int i = 3; i < responseLength + 3; i++){
                    printf("%c", buffer_udp[i]);
                }
                printf("\n");
            } else {                                                                    // Had to duplicate it because of the new line (ERR sends \n, OK does not)
                printf("OK:");
                for(int i = 3; i < responseLength + 3; i++){
                    printf("%c", buffer_udp[i]);
                }
                printf("\n");
            }
        }

    }

// TCP ------------------------------------------------------------------------------------------
    if(strcmp(conType, "tcp") == 0){      
    // SOCKET() - creating socket
        int type = SOCK_STREAM;
        int client_socket = socket(family, type, 0);
        if(client_socket <= 0){     
            fprintf(stderr, "ERROR: socket.\n");
            return 1;
        }

        //interruptSocket = client_socket;

    // CONNECT
        if(connect(client_socket, (struct sockaddr *) &server_address, sizeof(server_address)) != 0){
            fprintf(stderr, "ERROR: connect.\n");
            return 1;
        }

        while(keepRunning){
        // MESSAGE
            memset(buffer_tcp, 0, BUFFER_SIZE_TCP);                                     // Fills buffer with 0
            
            if(fgets(input_tcp, BUFFER_SIZE_TCP, stdin) == NULL){                       // Reads from input 
                fprintf(stderr, "ERROR occured while reading input.\n");
                if(!keepRunning){
                    strcpy(input_tcp, "BYE\n");
                    break;
                }
            }

            strcpy(buffer_tcp, input_tcp);                                              // Copies input into buffer
    
     // SEND()
            int bytes_tx = send(client_socket, buffer_tcp, strlen(buffer_tcp), 0);
            if(bytes_tx < 0){
                fprintf(stderr, "ERROR: send.\n");
            }

            memset(buffer_tcp, 0, BUFFER_SIZE_TCP);                                     // Fills buffer with 0

    // RECV()
            int bytes_rx = recv(client_socket, buffer_tcp, BUFFER_SIZE_TCP, 0);
            if(bytes_rx < 0){
                fprintf(stderr, "ERROR: recv.\n");
            }

            printf("%s", buffer_tcp);                                                   // Server's response

            if(strcmp(buffer_tcp, "BYE\n") == 0) break;
        }

    // CLOSE()
        if(!keepRunning){                                                               // If interrupted, this sends BYE and waits for response
            memset(buffer_tcp, 0, BUFFER_SIZE_TCP); 
            strcpy(buffer_tcp, input_tcp);

            int bytes_tx = send(client_socket, buffer_tcp, strlen(buffer_tcp), 0);
            if(bytes_tx < 0){
                fprintf(stderr, "ERROR: send.\n");
            }

            memset(buffer_tcp, 0, BUFFER_SIZE_TCP); 
            
            int bytes_sx = recv(client_socket, buffer_tcp, BUFFER_SIZE_TCP, 0);
            if(bytes_sx < 0){
                fprintf(stderr, "ERROR: recv.\n");
            }

            printf("%s", buffer_tcp);

        }
       
        shutdown(client_socket, SHUT_RD);
        shutdown(client_socket, SHUT_WR);
        shutdown(client_socket, SHUT_RDWR);

        close(client_socket);
    }

    return 0;
}


ipkcpc.c
// Project 1 - IPK calculator protocol
// Author: Milan Takac - xtakac09
// 2BIT VUT FIT (BUT)

#include <stdio.h>
#include <string.h>
#include <stdlib.h>         // atoi
#include <sys/socket.h>     // socket, sendto
#include <sys/types.h>
#include <netdb.h>          // gethostbyname
#include <arpa/inet.h>      // htons
#include <unistd.h>         // close
#include <signal.h>         // C-c interrupt signal

#define BUFFER_SIZE_UDP 512     
#define BUFFER_SIZE_TCP 1024     

static volatile int keepRunning = 1;                                                    // Help variable in case of interrupt

void intHandler() {
    keepRunning = 0;
    return;
}

int main(int argc, char** argv){
    signal(SIGINT, intHandler);                                                         // For C-c interrupt

// Parameter check 
    if(argc != 7){                                                                      // Number of args
        fprintf(stderr, "ERROR: Wrong number of arguments. %d of 6 needed.\n", argc);
        return 1;
    }

    if(strcmp(argv[1], "-h") != 0 && strcmp(argv[3], "-p") != 0 && strcmp(argv[5], "-m") != 0){ 
        fprintf(stderr, "ERROR: Wrong arguments. Correct form is \"ipkcpc -h <host> -p <port> -m <mode>\"\n");
        return 1;
    }

    if(!(strcmp(argv[6], "tcp") == 0 || strcmp(argv[6], "udp") == 0)){
        fprintf(stderr, "ERROR: Wrong mode \n%s\n - choose \"tcp\" or \"udp\".\n", argv[6]);
        return 1;
    }

    if(atoi(argv[4]) < 0){                                                              // Checks if port number isn't negative
        fprintf(stderr, "ERROR: Wrong port parameter: %s.\n", argv[4]);
        return 1;
    }

    char conType[4];                                                                    // Connection type (UDP/TCP)
    char host_address[16];               
    char buffer_udp[BUFFER_SIZE_UDP];
    char buffer_tcp[BUFFER_SIZE_TCP];
    char input_udp[BUFFER_SIZE_UDP];
    char input_tcp[BUFFER_SIZE_TCP];
    int family = AF_INET;                                                               // ipv4
    int port_number = atoi(argv[4]);

    strcpy(host_address, argv[2]);
    strcpy(conType, argv[6]);

    // Getting server info 
    struct hostent *server = gethostbyname(host_address);                               // Gets info about server. Parameter takes either address or name
    if(server == NULL){                                                                 // Invalid server name
        fprintf(stderr, "ERROR: no such host %s.\n", host_address);
        return 1;
    }

    struct sockaddr_in server_address;
    memset(&server_address, 0, sizeof(server_address));                                 // fills server address with zeros

    server_address.sin_family = AF_INET;
    server_address.sin_port = htons(port_number);                                       // sets port that the socket will use. htonl() translates an unsigned long integer into network byte order
    memcpy(&server_address.sin_addr.s_addr, server->h_addr_list[0], server->h_length); 

    printf("INFO: Server socket: %s : %d \n", 
            inet_ntoa(server_address.sin_addr),                                         // server's in address
            ntohs(server_address.sin_port));                                            // server's in port

    socklen_t server_size = sizeof(server_address);    

// UDP ------------------------------------------------------------------------------------------
    if(strcmp(conType, "udp") == 0){                                                    // UDP - no need to disconnect if an error occurs
        int type = SOCK_DGRAM;

    // SOCKET() - creating socket
        int client_socket = socket(family, type, 0);
        if(client_socket <= 0){     
            fprintf(stderr, "ERROR: socket.\n");
            return 1;
        }

        while(keepRunning){
        // MESSAGE
            memset(buffer_udp, 0, BUFFER_SIZE_UDP);                                     // Fills buffer with 0
            
            if(fgets(input_udp, BUFFER_SIZE_UDP, stdin) == NULL){                           // Reads from input_udp 
                fprintf(stderr, "ERROR occured while reading input.\n");
                return 1;
            }

            if(strcmp(input_udp, "exit\n") == 0){                                           // Exit message
                printf("Exiting...\n");
                return 0;
            }

            int inputLength = (int) strlen(input_udp) - 1;                                  // - 1, because '\n' is not considered as part of the payload length
            buffer_udp[0] = 0;                                                          // Opcode (0 = request)
            buffer_udp[1] = inputLength;                                                // Payload Length

            for(int i  = 0; i < inputLength; i++){                                      // Fills the payload data area with user input_udp
                buffer_udp[i + 2] = input_udp[i];                                      
            }

        // SENDTO() - sending message to server 
            int mistake = 0;
            int bytes_tx = sendto(client_socket, buffer_udp, (inputLength + 2), 0, (struct sockaddr *) &server_address, server_size);    // inputLength + 2, because opcode + payloadLength + input_udp (not counting '\0')
            if(bytes_tx < 0){
                fprintf(stderr, "ERROR: sendto.\n");
                mistake++;
            }

            if(mistake) continue;

            memset(buffer_udp, 0, sizeof(buffer_udp));                                  // Fills buffer with 0
            
        // RECVFROM() - waiting for response
            
            int bytes_rx = recvfrom(client_socket, buffer_udp, BUFFER_SIZE_UDP, 0, (struct sockaddr *) &server_address, &server_size);
            if(bytes_rx < 0){
                fprintf(stderr, "ERROR: recvfrom.\n");
            } 

            int responseStatus = buffer_udp[1];
            int responseLength = buffer_udp[3];
            if(responseStatus == 1){           // STATUS CODE: 0 = OK, 1 = ERR
                fprintf(stderr, "ERR:");
                for(int i = 3; i < responseLength + 3; i++){
                    printf("%c", buffer_udp[i]);
                }
                printf("\n");
            } else {                                                                    // Had to duplicate it because of the new line (ERR sends \n, OK does not)
                printf("OK:");
                for(int i = 3; i < responseLength + 3; i++){
                    printf("%c", buffer_udp[i]);
                }
                printf("\n");
            }
        }

        close(client_socket);
    }

// TCP ------------------------------------------------------------------------------------------
    if(strcmp(conType, "tcp") == 0){      
    // SOCKET() - creating socket
        int type = SOCK_STREAM;
        int client_socket = socket(family, type, 0);
        if(client_socket <= 0){     
            fprintf(stderr, "ERROR: socket.\n");
            return 1;
        }

    // CONNECT
        if(connect(client_socket, (struct sockaddr *) &server_address, sizeof(server_address)) != 0){
            fprintf(stderr, "ERROR: connect.\n");
            return 1;
        }

        while(keepRunning){
        // MESSAGE
            memset(buffer_tcp, 0, BUFFER_SIZE_TCP);                                     // Fills buffer with 0
            
            if(fgets(input_tcp, BUFFER_SIZE_TCP, stdin) == NULL){                       // Reads from input
                fprintf(stderr, "ERROR occured while reading input.\n");
                if(!keepRunning){
                    strcpy(input_tcp, "BYE\n");
                    break;
                }
            }

            strcpy(buffer_tcp, input_tcp);                                              // Copies input into buffer
    
     // SEND()
            int mistake = 0;
            int bytes_tx = send(client_socket, buffer_tcp, strlen(buffer_tcp), 0);
            if(bytes_tx < 0){
                fprintf(stderr, "ERROR: send.\n");
                mistake++;
            }

            if(mistake) continue;

            memset(buffer_tcp, 0, BUFFER_SIZE_TCP);                                     // Fills buffer with 0

    // RECV()
            int bytes_rx = recv(client_socket, buffer_tcp, BUFFER_SIZE_TCP, 0);
            if(bytes_rx < 0){
                fprintf(stderr, "ERROR: recv.\n");
            }

            printf("%s", buffer_tcp);                                                   // Server's response

            if(strcmp(buffer_tcp, "BYE\n") == 0) break;
        }

    // CLOSE()
        if(!keepRunning){                                                               // If interrupted, this sends BYE and waits for response
            memset(buffer_tcp, 0, BUFFER_SIZE_TCP); 
            strcpy(buffer_tcp, input_tcp);

            int bytes_tx = send(client_socket, buffer_tcp, strlen(buffer_tcp), 0);
            if(bytes_tx < 0){
                fprintf(stderr, "ERROR: send.\n");
            }

            memset(buffer_tcp, 0, BUFFER_SIZE_TCP); 
            
            int bytes_sx = recv(client_socket, buffer_tcp, BUFFER_SIZE_TCP, 0);
            if(bytes_sx < 0){
                fprintf(stderr, "ERROR: recv.\n");
            }

            printf("%s", buffer_tcp);

        }
       
        shutdown(client_socket, SHUT_RD);
        shutdown(client_socket, SHUT_WR);
        shutdown(client_socket, SHUT_RDWR);

        close(client_socket);
    }

    return 0;
}
